### PR TITLE
Issue #3407 aws_ssm_association handles updates incorrectly

### LINKED ISSUE #3407: aws_ssm_parameter with association name seems to interfere with parameters and vice versa
### Terraform Version
Terraform v0.11.3
+ provider.aws v1.8.0
+ provider.external v1.0.0
+ provider.template v1.0.0

### Affected Resource(s)
- aws_ssm_association

### Terraform Configuration Files
```hcl
resource "aws_ssm_association" "test" {
    association_name = "Test"
    name = "AWS-RunAnsiblePlaybook"
    targets = [
        {
            key = "tag:Test"
            values = ["true"]
        }
    ]
    parameters {
        check = "False"
        playbook = <<EOF
- hosts: all
  become: true
  tasks:
  - name: set debug header
    copy:
      dest: /etc/nginx/conf.d/test.conf
      content: add_header "X-Debug" "Test" always;
      mode: 0644
    notify:
    - reload nginx
  handlers:
  - name: reload nginx
    service:
      name: nginx
      state: reloaded
EOF
    }
}
```

### Expected Behavior
SSM Assocation should have been updated with new ansible script and assocaition name when updated

### Actual Behavior
This varies.

a) On the first run everything goes as expected.

b) If for example the ansible script gets changed, and the terraform is re-run, the parameters will be completely empty

c) If run again (no further changes) then terraform will still detect a change, then do the update, the parameters will be there but the association name will be '-'

d) If run again (no further changes) then terraform will still detect a change, then do the update, the parameters will be empty again, but now the association name will be correct.

e) goto c

### PR DESCRIPTION
**Pull request by dpldb**
Fixes #3407

Changes proposed in this pull request:

* Fixed Typo: Corrected Assoscation to Assocation
* resourceAwsSsmAssocationUpdate needs to pass everything to AWS because the AWS API creates a new version of the association rather than updating it in-place. I updated the function to check for any changes and update everything if there are.

Output from acceptance testing:

```
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSSSMAssociation'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -run=TestAccAWSSSMAssociation -timeout 120m
=== RUN   TestAccAWSSSMAssociation_basic
--- PASS: TestAccAWSSSMAssociation_basic (146.89s)
=== RUN   TestAccAWSSSMAssociation_withTargets
--- PASS: TestAccAWSSSMAssociation_withTargets (38.64s)
=== RUN   TestAccAWSSSMAssociation_withParameters
--- PASS: TestAccAWSSSMAssociation_withParameters (26.51s)
=== RUN   TestAccAWSSSMAssociation_withAssociationName
--- PASS: TestAccAWSSSMAssociation_withAssociationName (25.49s)
=== RUN   TestAccAWSSSMAssociation_withDocumentVersion
--- PASS: TestAccAWSSSMAssociation_withDocumentVersion (16.49s)
=== RUN   TestAccAWSSSMAssociation_withOutputLocation
--- PASS: TestAccAWSSSMAssociation_withOutputLocation (68.60s)
=== RUN   TestAccAWSSSMAssociation_withScheduleExpression
--- PASS: TestAccAWSSSMAssociation_withScheduleExpression (25.53s)
PASS
ok  	github.com/dpldb/terraform-provider-aws/aws	348.188s
...
```

### CODE REVIEW

**Comment by ajjl** [comment]
I am encountering the same issue. Thanks for addressing this.

File: aws/resource_aws_ssm_association.go [hunk 1]
```diff
@@ -186,37 +186,67 @@ func resourceAwsSsmAssociationRead(d *schema.ResourceData, meta interface{}) err
...
-	log.Printf("[DEBUG] SSM association update: %s", d.Id())
+	log.Printf("[DEBUG] SSM Association update: %s", d.Id())
 
 	associationInput := &ssm.UpdateAssociationInput{
 		AssociationId: aws.String(d.Get("association_id").(string)),
 	}
 
+	// AWS creates a new version every time the association is updated, so everything should be passed in the update.
+
+	hasChanges := false
```
Comment (bflad) [inline aws/resource_aws_ssm_association.go]: The `resourceAwsSsmAssociationUpdate` function is not performing any other API calls beyond `UpdateAssociation` so all the `hasChanges` logic can be removed 👍

Comment (dpldb) [reply aws/resource_aws_ssm_association.go]: Good point. I'll get rid of this.

**Comment by dpldb** [comment]
@bflad Thanks for the code review and quick response. Happy to help. I'll take a look at the acceptance test.

**Comment by dpldb** [comment]
Thanks @bflad

**Comment by bflad** [comment]
This has been released in [version 1.34.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1340-august-30-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Pull request merged**